types/jest/index.d.ts

```diff
@@ -762,12 +763,12 @@ declare namespace jest {
         new (...args: any[]): any;
     }
 
-    interface Mock<T = {}> extends Function, MockInstance<T> {
-        new (...args: any[]): T;
-        (...args: any[]): any;
+    interface Mock<T, Y extends any[]> extends Function, MockInstance<T, Y> {
```

Comment:
Please restore the default type parameter and add one for Y. I don't think making these required is worth the breaking change, and trying to "force people" to use stronger types in tests is very opinionated. A large percentage of `jest.Mock` usages don't touch values, only call counts. Also keep in mind the inference for stronger mock types will only help with spyOn. Every other declaration will need to have the parameters passed explicitly, which is going to be a bit painful since Y is an arguments tuple that needs to be extracted.

Reply:
Please we want to move away from `any` when we are with typescript => so no default value with `any`

Arguments:
1.
We need this contract for robust integrated test. I think this had to be corrected long ago (ts 2.8)
in other case the test are reall weak.


2.
This fix will make jasmine behind compared to jest and will raise the adoption of the library.

3.
https://palantir.github.io/tslint/rules/no-any/
`Using any as a type declaration nullifies the compile-time benefits of the type system.`

@rickhanlonii can we have your opinion here, please?

Reply:
> A large percentage of jest.Mock usages don't touch values, only call counts

@jwbay can you explain this with an example of what you can do now and what this change would require?

Reply:
> I know that is it very opinionated, but shouldn't we, as a community, enforce best practices?

To a point. We need to balance the benefits and costs considering the pain of breaking changes. As a data point, it's possible to make a package so annoyingly type-safe the change is reverted: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26813

> I strongly believe that testing should be done the safest way possible, and "any" type should be avoided at all cost.

You are absolutely entitled to that strong opinion in your projects. The issue here is that this change impacts a large number of developers across the world, from enterprise teams to personal projects (this package gets 2 million downloads per week), for value that people will question. This change as-is isn't really just opinionated—it's prescriptive.

> I don't agree that stronger mock types will only help for spyOn. There is a couple of example I changed in jest-tests (mockContextVoid, mockContextString, spy3Mock, spy4) where the typing of arguments would be lost and potentially mislead people.

That's fair, I didn't qualify my statement correctly. spyOn is the only way to infer types *from existing functions*. Everything else will either need explicit types, a runtime function to extract them via inference, or some kind of conditional type to convert `typeof MyFunction` to a correctly populated `jest.Mock`. We should consider adding that last option with this change, actually, assuming it goes forward. Looks like it's just `T extends (...args: infer R) => infer V ? jest.Mock<V, R> : never`

> If you don't care about tuple generics, you can still use <any, any[]> or without typing :

That's part of the problem with this. The <any, any[]> is often just noise. Reading noise sucks and writing noise really sucks. Here are a couple examples:

```tsx
test('foo', () => {
  someFunction();
  // type params are just noise here
  (foo as jest.Mock<any, any[]>).mockClear();
  someFunction();
  expect(fooMock).toHaveBeenCalledTimes(1);
});

test('foo', () => {
  // more noise
  (foo as jest.Mock<any, any[]>).mockName('MyMock');
  const rendered = render(<Component foo={foo} />);
  expect(rendered).toMatchSnapshot();
});
```

I think providing defaults for the cases where the type parameters truly don't matter is a reasonable ask. How about a compromise of defaulting to `never`? That way you can't interact with arguments or return values unsafely by default, and you can't supply an invalid implementation. It would still be a massive breaking change, but at least the types involved are being touched in some way, so it's more reasonable as a value proposition.

I'd really like to get the thoughts of other maintainers here considering the impact: @NoHomey, @asvetliakov, @alexjoverm, @epicallan, @ikatyang, @wsmd, @JamieMason, @douglasduteil, @ahnpnl, @joshuakgoldberg, @UselessPickles, @r3nya, @hotell, @sebald, @andys8

As a note for impact, things like this are going to start popping up in codebases consuming this change:

```ts
// this function only touches one or two properties of a larger interface
mockFoo.mockReturnValue({
  x: 42
} as Partial<RealReturnType> as RealReturnType);

mockFoo.mockImplementation((a, _b, _c, _iReallyDontCare, _makeItStop) => {
  return Promise.resolve(a)
});
```

Reply:
Sorry, it's a little hard to parse through all the discussion and changes - what exactly are the breaking changes? Where are folks now being forced to use types when things defaulted to `any` before?

It sounds like they might be significant _(for better or for worse)_, so even if they're the correct thing to do, there should be a clear summary of what's different. Jest is a popular enough library that any breaking change, no matter how small, is likely confusing and painful to some users.

Reply:
updated description with breaking change

Reply:
Thanks!

Although the rest of the changes look _excellent_ (really looking forward to being able to strictly type spies/mocks!), that one point of _requiring_ explicit type definitions on variables is not a good thing for many users. IMO we should default to `any` when not provided, not `never`.

Agreed that it's often better practice to explicitly type things, but not all users are interested in best practices, or the same best practices. Some users are just interested in having tests work out of the box, and any added pain given to them by TypeScript is a negative - even if it gives better type checking. A few examples:
* Large codebases converting from JavaScript would want as few places to _need_ to explicitly type variables as possible.
* In a lot of folks' minds, unit tests don't need nearly the same level of type checking as source code, and if you explicitly fail/error within a few milliseconds, needing to write & update them constantly just gets in the way.

Speaking more generally: what are "best practices"? If your goals are to have completely strict type safety, then sure, enforcing explicit types helps enforce your personal best practices. But not everybody has that same goal. Some just want to write code quickly, be able to define a few non-object constructs such as interfaces, and get some improved level of IDE features. IMO we shouldn't sacrifice one use case (stricter typing) for others (quicker development, easier TS migration).

Reply:
I understand all your arguments and i think it's case by case, team by team, project by project. A 
- Migration may require at the beginning an `any` default value. 
- Experienced people may not need to force them do it right when at the opposite other team needs it. 

So i am calling a crazy idea. 
How to make this opinionated default value **parametrizable ** for each of us depending on the needs.

Reply:
ok! i am trying the flags to see if we can save the situation :) it by config
But i did what test that shows me a pure inconsistency
![image](https://user-images.githubusercontent.com/1953635/49688020-047e4a00-fada-11e8-9aa0-0b7cfc9c72ec.png)

if the plan is to add only defaults value `any` for jest.Mock and jest.SpyInstance>
The compiler complaining for the second test but not the first one => `inconsistency`

Reply:
`noImplicitAny` didn't work.  

 if default `any` used in jest type:
I found a way by wrapping the type
`type CustomSpyInstance<A = never> = jest.SpyInstance<A>;` which override `any`
with this it will prevent devs from passing wrong type and this can be done at the project level not impacting all the devs around the world.

one detail:
To get the right type for case1 from example on top i had to:
`let spy: CustomSpyInstance<HeroService['computeMaxLevel']>;`
In other case i had a type error when i use the `spyOn` i couldnt solve without this trick.

Reply:
noImplicitAny didn't work.

if default values (any) used in jest types:
I found a way by wrapping the type
```type CustomSpyInstance<A = never, B = never> = jest.SpyInstance<A, B>;``` which override any
with this it will prevent devs from passing wrong type and this can be done at the project level not impacting all the devs around the world (migrating from js to ts ;) )